### ISSUE TITLE
Failed to discover nodes as the discovery method is set to undef by wrong

### ISSUE THREAD

**Issue by robin2008** [xcat2/xcat-core#4860]
In one customer site, they nodes discovering slow down after many racks are done. From the logs, we can see the node has been handled in switch-discovery mode without error, but finally, it reports "Failed for node discovery."

In our test environment, we also find similar issue. 
```
nodediscoverls
Discovered 16 node.
  UUID                                    NODE                METHOD         MTM        SERIAL
  ****                                      sn01                  switch         8335-GTB     ****
  ****                           mid05tor12cn16           undef         8335-GTB     ****
```
Actually the mid05tor12cn16 is discovered, but its method is `undef`. And from the logs we can see
```
Feb 28 03:01:11 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Beginning node discovery process...
Feb 28 03:01:11 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Waiting for nics to get addresses
Feb 28 03:01:12 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Network configuration complete, commencing transmit of discovery packets
Feb 28 03:01:12 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Beginning echo information to discovery packet file...
Feb 28 03:01:16 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Discovery packet file is ready.
Feb 28 03:01:16 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Sending the discovery packet to xCAT (briggs01:3001)...
Feb 28 03:01:16 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Sleeping 5 seconds...
Feb 28 03:01:15 briggs01 xcat[113087]: xcatd: Processing discovery request from 172.12.139.16
Feb 28 03:01:15 briggs01 xcat[113087]: xcat.discovery.aaadiscovery: (AA:BB:CC:DD:EE:FF) Got a discovery request, attempting to discover the node...
Feb 28 03:01:15 briggs01 xcat[113087]: xcat.discovery.blade: (AA:BB:CC:DD:EE:FF) Warning: Could not find any nodes using blade-based discovery
Feb 28 03:01:18 briggs01 xcat[113087]: refresh_table ElapsedTime:3 sec
Feb 28 03:01:18 briggs01 xcat[113087]: xcat.discovery.switch: (AA:BB:CC:DD:EE:FF) Found node: mid05tor12cn16
Feb 28 03:01:18 briggs01 xcat[113087]: xcat.discovery.nodediscover: Removing discovered node definition: ...
Feb 28 03:01:19 briggs01 xcat[120238]: xCAT: Allowing rmdef for root from localhost
Feb 28 03:01:19 briggs01 xcat[113087]: xcat.discovery.nodediscover:  definition removed from xCAT
Feb 28 03:01:19 briggs01 xcat[113087]: xcat.discovery.nodediscover: mid05tor12cn16 has been discovered
Feb 28 03:01:19 briggs01 xcat[113087]: xcat.discovery.zzzdiscovery: (AA:BB:CC:DD:EE:FF) Failed for node discovery.
Feb 28 03:01:19 briggs01 xcat[113087]: xcat.discovery.zzzdiscovery: Notify 172.12.139.16 that its findme request has been processed
Feb 28 03:01:21 mid05tor12cn16 [localhost] xcat.genesis.dodiscovery: Restart network interfaces...
Feb 28 03:01:21 mid05tor12cn16 [localhost] dhclient[24748]: DHCPRELEASE on enP5p1s0f0 to 172.12.253.27 port 67 (xid=0x4885f2d1)
Feb 28 03:01:20 briggs01 dhcpd: DHCPRELEASE of 172.12.139.16 from AA:BB:CC:DD:EE:FF via enP34p1s0f0.12 (not found)
Feb 28 03:01:20 briggs01 dhcpd: DHCPDISCOVER from AA:BB:CC:DD:EE:FF via enP34p1s0f0.12
Feb 28 03:01:20 briggs01 dhcpd: DHCPOFFER on 172.12.139.16 to AA:BB:CC:DD:EE:FF via enP34p1s0f0.12
Feb 28 03:01:20 briggs01 dhcpd: DHCPREQUEST for 172.12.139.16 (172.12.253.27) from AA:BB:CC:DD:EE:FF via enP34p1s0f0.12
Feb 28 03:01:20 briggs01 dhcpd: DHCPACK on 172.12.139.16 to AA:BB:CC:DD:EE:FF via enP34p1s0f0.12
```

**Comment by robin2008** [comment]
After code debug and add logs, it is found that the `undef` is set by `zvmdiscovery.pm`
```
    unless ( $ZVMdiscover ) {
        if ( $SEQdiscover or $PCMdiscover ) {
            # profile or sequential discovery is running, then just return 
            # to make the other discovery handle it
            return;
        }
        # update the discoverydata table to have an undefined node
        $request->{discoverymethod}->[0] = 'undef';                         <========   this line cause the issue
        xCAT::DiscoveryUtils->update_discovery_data($request);
        return;
    }
```

**Comment by robin2008** [comment]
@chuckbrazie 
could you take a look at `zvmdiscovery.pm`?  I didn't understand the above logic.
It seems when `$ZVMdiscover` is set, it do nothing. And when not set, your plugins will run something, especially to update DB, why?

**Comment by robin2008** [comment]
the issue will be addressed in PR #4863
